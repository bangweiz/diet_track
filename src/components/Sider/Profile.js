import React from "react";
import PropTypes from "prop-types";

import Avart from "../../asset/avatar.png";
import "./Profile.scss";

const Profile = props => {
  return (
    <div className="profile">
      <div className="profile-weight">
        <span className="profile-number">{props.weight_kg}</span>
        <span className="profile-unit">kg</span>
      </div>
      <img
        className="profile-avatar"
        alt="avatar"
        src={Avart}
      ></img>
      <div className="profile-height">
        <span className="profile-number">{props.height_cm}</span>
        <span className="profile-unit">cm</span>
      </div>
      <span className="profile-name">
        {props.first_name}&nbsp;
        <span className="profile-last-name">{props.last_name}</span>
      </span>
    </div>
  );
};

Header.propTypes = {
  first_name: PropTypes.string.isRequired,
  last_name: PropTypes.string.isRequired,
  weight_kg: PropTypes.number.isRequired,
  height_cm: PropTypes.number.isRequired
}

export default Profile;
