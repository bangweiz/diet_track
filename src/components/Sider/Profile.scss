.profile {
  width: 100%;
  text-align: center;
  position: relative;
  height: 192px;
  border-bottom: 1px solid rgba(0, 0, 0, .12);

  @media screen and (max-width: 768px) {
    display: none;
    border: none;
  }

  .profile-weight, .profile-height {
    width: 56px;
    height: 56px;
    background-color: rgba($color: #000000, $alpha: .5);
    color: #fff;
    border-radius: 50%;
    position: absolute;
    top: 44px;

    .profile-number {
      display: inline-block;
      width: 40px;
      height: 24px;
      line-height: 24px;
      font-size: 20px;
      margin: 8px;
      margin-bottom: 0;
      letter-spacing: .25px;
      font-weight: 500;
    }

    .profile-unit {
      display: inline-block;
      width: 40px;
      height: 16px;
      font-size: 12px;
      line-height: 1.33;
      letter-spacing: .4px;
      transform: translateY(-5px);
    }
  }

  .profile-weight {
    right: 232px;
  }

  .profile-height {
    right: 32px;
  }

  .profile-avatar {
    width: 96px;
    height: 96px;
    border-radius: 50%;
    position: absolute;
    top: 24px;
    left: 112px;
    background-color: rgba($color: #000000, $alpha: .5);
  }

  .profile-name {
    display: inline-block;
    width: 288px;
    height: 28px;
    line-height: 28px;
    font-size: 24px;
    color: rgba(0, 0, 0, 0.87);
    position: absolute;
    top: 132px;
    left: 16px;

    .profile-last-name {
      @media screen and (max-width: 768px) {
        display: none;
      }
    }
  }
}