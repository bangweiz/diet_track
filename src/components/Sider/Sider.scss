.sider {
  width: 320px;
  background-color: #f5f5f5;
  flex-shrink: 0;
  overflow: auto;

  @media screen and (max-width: 768px) {
    width: 100%; 
  }
}