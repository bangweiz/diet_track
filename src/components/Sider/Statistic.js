import React, { PureComponent } from "react";
import { connect } from "react-redux";

import "./Statistic.scss";

class Statistic extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      totalCal: 0,
      breakfast: 0,
      lunch: 0,
      dinner: 0,
      snack: 0,
      percentage: 0
    };
    this.barRef = React.createRef();
    this.perRef = React.createRef();
  }

  componentDidUpdate(prevProp) {
    if (this.props !== prevProp) {
      if (this.props.data.length === 0) {
        this.calculate(this.props.savedFood);
      } else {
        this.calculate(this.props.data);
      }
    }
  }

  render() {
    return (
      <div className="statistic">
        <div className="statistic-cal">
          <div className="statistic-cal-consumed">
            <span className="number">{this.state.totalCal} Cal</span>
            <span className="description">consumed</span>
          </div>
          <div className="statistic-cal-goal">
            <span className="number">{this.props.daily_goal} Cal</span>
            <span className="description">daily goal</span>
          </div>
        </div>
        <div className="statistic-percentage">
          <span className="progress-bar">
            <span ref={this.barRef} className="progress-bar-before"></span>
          </span>
          <span ref={this.perRef} className="number">{this.state.percentage}%</span>
        </div>
        <div className="statistic-meal">
          <div className="statistic-meal-common">
            <span className="number">{this.state.breakfast}</span>
            <span className="meal">Breakfast</span>
          </div>
          <div className="statistic-meal-common">
            <span className="number">{this.state.lunch}</span>
            <span className="meal">Lunch</span>
          </div>
          <div className="statistic-meal-common">
            <span className="number">{this.state.dinner}</span>
            <span className="meal">Dinner</span>
          </div>
          <div className="statistic-meal-common">
            <span className="number">{this.state.snack}</span>
            <span className="meal">Snack</span>
          </div>
        </div>
      </div>
    );
  }

  calculate = items => {
    let totalCal = 0;
    let breakfast = 0;
    let lunch = 0;
    let dinner = 0;
    let snack = 0;
    let percentage = 0;

    items.forEach(item => {
      switch (item.meal_type) {
        case "breakfast": {
          breakfast += item.nf_calories;
          totalCal += item.nf_calories;
          break;
        }
        case "lunch": {
          lunch += item.nf_calories;
          totalCal += item.nf_calories;
          break;
        }
        case "dinner": {
          dinner += item.nf_calories;
          totalCal += item.nf_calories;
          break;
        }
        case "snack": {
          snack += item.nf_calories;
          totalCal += item.nf_calories;
          break;
        }
        default: {
          break;
        }
      }
    });
    percentage = Math.floor((totalCal / this.props.daily_goal) * 100);
    this.barRef.current.style.width = percentage > 100 ? '100%' : (percentage + '%');
    this.perRef.current.style.marginLeft = `${(2.88 * percentage) > 288 ? 280 : (2.88 * percentage)}px`;
    this.setState({ totalCal, breakfast, lunch, dinner, snack, percentage });
  }
}

const mapStateToProps = state => ({
  savedFood: state.savedFood
});

export default connect(
  mapStateToProps,
  {}
)(Statistic);
