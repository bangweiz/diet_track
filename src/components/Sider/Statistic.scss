.statistic {
  width: 100%;
  padding: 24px 16px;

  .statistic-cal {
    @media screen and (max-width: 768px) {
      display: flex;
      justify-content: space-around;
    }

    .statistic-cal-consumed, .statistic-cal-goal {
      display: inline-block;
      width: 144px;

      .number {
        display: inline-block;
        width: 100%;
        height: 28px;
        font-size: 24px;
        color: rgba(0, 0, 0, 0.87);
      }

      .description{
        display: inline-block;
        width: 100%;
        height: 16px;
        font-family: Roboto;
        font-size: 12px;
        line-height: 1.33;
        letter-spacing: 0.4px;
        color: rgba(0, 0, 0, 0.6);
      }
    }

    .statistic-cal-goal {
      text-align: right;
    }
  }

  .statistic-percentage {
    width: 100%;
    margin-top: 10px;
    margin-bottom: 20px;
    
    .progress-bar {
      display: inline-block;
      width: 100%;
      height: 4px;
      background-color: #d3bbf3;
      position: relative;

      .progress-bar-before {
        width: 0;
        height: 4px;
        background-color: #6200ee;
        position: absolute;
        top: 0;
        left: 0;
      }
    }

    .number {
      display: inline-block;
      height: 24px;
      font-size: 14px;
      font-weight: 500;
      line-height: 1.71;
      letter-spacing: 0.1px;
      text-align: center;
      color: rgba(0, 0, 0, 0.6);
      transform: translateX(-50%);
    }
  }
  
  .statistic-meal {
    width: 100%;

    @media screen and (max-width: 768px) {
      display: flex;
      justify-content: space-around;
    }

    .statistic-meal-common {
      width: 66px;
      display: inline-block;

      &:not(:last-child) {
        margin-right: 8px;
      }

      .number {
        display: inline-block;
        width: 100%;
        height: 24px;
        font-size: 20px;
        font-weight: 500;
        letter-spacing: 0.25px;
        text-align: center;
        color: rgba(0, 0, 0, 0.87);
      }

      .meal {
        display: inline-block;
        width: 100%;
        height: 16px;
        font-size: 12px;
        line-height: 1.33;
        letter-spacing: 0.4px;
        text-align: center;
        color: rgba(0, 0, 0, 0.6);
      }
    }
  }
}