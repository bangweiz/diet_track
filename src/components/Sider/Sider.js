import React, { PureComponent } from "react";

import Profile from "./Profile";
import Statistic from "./Statistic";
import "./Sider.scss";

class Sider extends PureComponent {
  render() {
    return (
      <div className="sider">
        <Profile
          first_name={this.props.first_name}
          last_name={this.props.last_name}
          weight_kg={this.props.weight_kg}
          height_cm={this.props.height_cm}
        />
        <Statistic data={this.props.data} daily_goal={this.props.daily_goal} />
      </div>
    );
  }
}

export default Sider;
