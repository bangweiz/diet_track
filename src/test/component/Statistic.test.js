import React from "react";
import Enzyme, { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

import { Statistic } from "../../components/Sider/Statistic";

Enzyme.configure({
  adapter: new Adapter()
});

test("Should render Statistic page", () => {
  const wrapper = shallow(<Statistic savedFood={{}} />);
  expect(wrapper.find('.statistic').length).toBe(1);
});