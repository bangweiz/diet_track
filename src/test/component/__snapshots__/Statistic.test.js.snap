// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Statistic page 1`] = `ShallowWrapper {}`;
